torchvision/datasets/triplet.py

```diff
@@ -0,0 +1,85 @@
+import torch
+import torch.utils.data as data
+
+import math
+from collections import defaultdict
+
+
+def get_class_samples(samples):
+    """Bins samples with respect to classes
+
+    Args:
+        num_classes (int): number of classes
+        samples (list[tuple]): (sample, class_idx) tuples
+
+    Returns:
+        list[list]: Bins of sample paths, binned by class_idx
+    """
+    class_samples = defaultdict(list)
+    for sample_path, class_idx in samples:
+        class_samples[class_idx].append(sample_path)
+
+    return class_samples
+
+
+def generate_triplet(class_samples):
+    """Generates a triplet from bins of samples
+
+    Args:
+        class_samples (list[list]): bins of samples, binned by class
+        num_triplets (int): number of triplets to be generated
+
+    Returns:
+        list[tuple]: triplets of the form (anchor, positive, negative)
+    """
+    pos_cls, neg_cls = torch.multinomial(torch.ones(len(class_samples)), 2).tolist()
+    pos_samples, neg_samples = class_samples[pos_cls], class_samples[neg_cls]
+
+    anc_idx, pos_idx = torch.multinomial(torch.ones(len(pos_samples)), 2).tolist()
+    neg_idx = torch.multinomial(torch.ones(len(neg_samples)), 1).item()
+
+    return (pos_samples[anc_idx], pos_samples[pos_idx], neg_samples[neg_idx])
+
+
+class TripletDataset(data.IterableDataset):
+    """
+    A dataset with samples of the form (anchor, positive, negative), where anchor and
+    positive are samples of the same class, and negative is a sample of another class.
+    TripletDataset reads fram Dataset `dset` where `dset[i]` returns (sample_path, class_idx).
+
+    Args:
+        dset (Dataset): Dataset object where __getitem__ returns (sample_path, class_idx) tuple.
+        loader (callable): A function to load a sample given its path.
+        num_triplets (int): Number of triplets to generate before raising StopIteration.
+        transform (callable, optional): A function/transform that takes in
+            a sample and returns a transformed version.
+            E.g, ``transforms.RandomCrop`` for images.
+
+    Attributes:
+        samples (list[tuple]): List of (anchor, positive, negative) triplets
+    """
+
+    def __init__(self, dset, loader, num_triplets, transform=None):
```

Comment:
From the discussion in https://github.com/pytorch/vision/issues/1042#issuecomment-506979156, I think that if we want to let this support hard-negative mining, we should go away from only having the class information, but we should instead have a similarity matrix that is also passed for example.

In order to be future-proof, what about instead of specifying a `loader` and a `dset`, we only pass a `dataset` (which can be arbitrary, just need to be index-based), and an extra list of groups (which have the size of the the dataset), we could then extend this later to other more complex use-cases.

So a possible API:
```python
def __init__(self, dataset, num_triplets, groups, transform=None):
    assert len(groups) == len(dataset)
```

Reply:
This does look a lot cleaner. To clarify, `groups[i]` will be the group that the sample `dataset[i]` belongs to, right? Also, will the group type be restricted to `int` or can it be any type?

Reply:
> groups[i] will be the group id that the sample dataset[i] is similar to

yes, that's right, and is similar to how I used it in `GroupedBatchSampler` in https://github.com/pytorch/vision/blob/867e5b4e7835dbebcf0c501850a7b08e466496f1/references/detection/group_by_aspect_ratio.py#L29

> Also, will the group type be restricted to int or can it be any dictionary indexable type?

I'd say that any dict indexable type should be fine, but it restricts a bit the implementations that we might have under the hood. What use-cases would you think that something other than an `int` would be useful?

Reply:
None really (I would just use a `group_to_idx` dict if we wanted to support other types). I was just clarifying my thoughts from an implementation perspective. Keeping it to `int` only would simplify implementation as well. Will keeping it limited to `int` be fine?

Reply:
Also, we won't require the `transform=None` kwarg anymore, right? I'm assuming the loading and transform will be taken care of by the input dataset's `__getitem__` method.

Reply:
@dwarakrajagopal 
> Will keeping it limited to int be fine?

yes, let's keep it as an int for now

> Also, we won't require the transform=None kwarg anymore, right?

Hum, what if the user wants to apply a transform to the whole triplet? I'm not sure there is a need for it, just thinking about a possibility

Reply:
Yeah I encountered such a problem while working on it. While testing, I gave the TripletDataset an ImageFolder dataset as input, and the triplet output I was getting was in this form: `((img0, cls0), (img1, cls0), (img2, cls1))`. A transform on the triplet would have been able to fix this to `(img0, img1, img2)`. I'll keep that in mind while working on it.

Reply:
Also note that having the `((img0, cls0), (img1, cls0), (img2, cls1))` is a reasonable output, because users might want to apply different losses / weights depending on the classes maybe?
Anyway, I think a transform is a good thing to have in this case.

Reply:
Yes it is reasonable, which is why I agreed that keeping transform was a good idea. I have made all these changes.